### DISCUSSION TITLE
whether to consider the rotation of grains when loading in the crystal plasticity model

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#25849] [Q&A Modules: Solid mechanics] [answered]
Dear @sapitts and @jiangwen84,

I hope this message finds you well. I have been studying the code in `ComputeElasticityTensorCP`, specifically in code block, 

```c++
void
ComputeElasticityTensorCP::computeQpElasticityTensor()
{
  // Properties assigned at the beginning of every call to material calculation
  // is required by the monolithic and user object versions. If those classes
  // are deprecated, these update can be removed and save time
  if (!_user_provided_rotation_matrix)
  {
    assignEulerAngles();
    _crysrot[_qp] = _R.transpose();
  }
  else
    _crysrot[_qp] = _rotation_matrix.transpose();

  _elasticity_tensor[_qp] = _Cijkl;
  _elasticity_tensor[_qp].rotate(_crysrot[_qp]);
}
```
and it appears that the `_elasticity_tensor[_qp]` is computed based on the initial input Euler angles, which represent the rotation matrix. In other words, it seems that throughout the computation, the `_crysrot[_qp]` remains constant. Could you please confirm if my understanding is accurate?

I am interested in knowing if grains' rotation is considered during the application of loading in crystal plasticity models.

Thank you for your time and assistance.

Best regards,

wei peng

**Comment by sapitts** [comment, marked as answer]
Hi Wei Peng,

Please accept my apologies for my late reply. In case you haven't already worked through this issue, the grain's rotation is not used in the calculation of crystal plasticity values. The stress and strain measures for the crystal plasticity capability (the PK2 stress and the Almansi-Hemel strain measures, respectively) are chosen to refer to the reference configuration. Because the crystal plasticity calculations are performed in the reference configuration, the updated rotation is not needed and thus is not calculated in the elasticity tensor class.

The updated rotation is, however, calculated and tracked by the `ComputeMultipleCrystalPlasticityStress` class, see https://github.com/idaholab/moose/blob/6ce5723f2e7b27ae5920296ae4093a7d77a18f0d/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L312  This updated rotation can also be tracked and output by the [ComputeUpdatedEulerAngle](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeUpdatedEulerAngle.html) class. 

Hope this helps,
Stephanie

**Reply by PengWei97** [reply]
No problem at all! Thank you for your answer. I got it.
